modules/peridynamics/src/actions/MechanicsActionPD.C

```diff
@@ -0,0 +1,220 @@
+                                              "peridynamic mechanics kernel will be applied to");
+  params.addParam<std::vector<AuxVariableName>>("save_in", "The displacement residuals");
+  params.addParam<std::vector<AuxVariableName>>("diag_save_in",
+                                                "The displacement diagonal preconditioner terms");
+  params.addParam<std::vector<MaterialPropertyName>>(
+      "eigenstrain_names",
+      "List of eigenstrains to be coupled in non-ordinary state-based mechanics kernels");
+
+  return params;
+}
+
+MechanicsActionPD::MechanicsActionPD(const InputParameters & params)
+  : Action(params),
+    _displacements(getParam<std::vector<VariableName>>("displacements")),
+    _ndisp(_displacements.size()),
+    _formulation(getParam<MooseEnum>("formulation")),
+    _stabilization(getParam<MooseEnum>("stabilization")),
+    _finite_strain_formulation(getParam<bool>("finite_strain_formulation")),
+    _save_in(getParam<std::vector<AuxVariableName>>("save_in")),
+    _diag_save_in(getParam<std::vector<AuxVariableName>>("diag_save_in"))
+{
+  // Consistency check
+  if (_save_in.size() != 0 && _save_in.size() != _ndisp)
+    mooseError("Number of save_in variables should equal to the number of displacement variables ",
+               _ndisp);
+
+  if (_diag_save_in.size() != 0 && _diag_save_in.size() != _ndisp)
+    mooseError(
+        "Number of diag_save_in variables should equal to the number of displacement variables ",
+        _ndisp);
+}
+
+void
+MechanicsActionPD::act()
+{
+  if (_current_task == "add_aux_variable")
+  {
+    // add the bond_status aux variable which is not restricted to any specific block but for
+    // all peridynamic domains
+    _problem->addAuxVariable("bond_status",
+                             FEType(Utility::string_to_enum<Order>("CONSTANT"),
+                                    Utility::string_to_enum<FEFamily>("MONOMIAL")));
+
+    // Set the initial value to unit using InitialConditionAction, following the coding in
+    // createInitialConditionAction() in AddVariableAction.C
+
+    // Set the parameters for the action
+    InputParameters action_params = _action_factory.getValidParams("AddOutputAction");
+    action_params.set<ActionWarehouse *>("awh") = &_awh;
+    action_params.set<std::string>("type") = "ConstantIC";
+    // Create the action
+    std::shared_ptr<MooseObjectAction> action = std::static_pointer_cast<MooseObjectAction>(
+        _action_factory.create("AddInitialConditionAction", "bond_status_moose", action_params));
+    // Set the required parameters for the object to be created
+    action->getObjectParams().set<VariableName>("variable") = "bond_status";
+    action->getObjectParams().set<Real>("value") = 1.0;
+    // Store the action in the ActionWarehouse
+    _awh.addActionBlock(action);
+  }
+  else if (_current_task == "setup_quadrature") // set the quadrature type to GAUSS_LOBATTO and
```

Comment:
Let's investigate whether this is actually necessary.  I'd rather not set this to prepare us for running together with finite elements, which might have higher order quadrature.